src/vdom/diff.js

```diff
@@ -287,7 +287,10 @@ export function recollectNodeTree(node, unmountOnly) {
 	else {
 		// If the node's VNode had a ref function, invoke it with null here.
 		// (this is part of the React spec, and smart for unsetting references)
-		if (node[ATTR_KEY]!=null) applyRef(node[ATTR_KEY].ref, null);
+		if (node[ATTR_KEY]!=null) {
+			applyRef(node[ATTR_KEY].ref, null);
+			delete node[ATTR_KEY].ref;
```

Comment:
This just always calls refs. I think we need to either hoist ref invocation back out of `setProperty()` (it used to happen during rendering), or wait for component recycling to go away.